### PR TITLE
Speed up sparklines when larger periods are selected and improve detection of trends as well as fix comparison of dates

### LINKED ISSUE #20066: Change sparklines visualisation depending on the selected date range
Currently, when viewing a sparkline over a longer data range, then we're visualising the data of every single day in that range. When you fetch a period of one year, it means Matomo has to select 365 different archives and plot each single dot on the sparkline.

[Example:](https://demo.matomo.cloud/index.php?module=CoreHome&action=index&idSite=1&period=day&date=yesterday#?idSite=1&period=day&date=yesterday&category=General_Visitors&subcategory=General_Overview)
image

To improve performance, and also to see trends better, we want to change the period used to plot the sparkline depending on the selected date range.

**Expected outcomes:**

* If a range of 24+ months is selected, then in the UI show the sparklines for the selected range for months.
 * For example, if a date range of `2018-06-21,2022-04-19` is selected, then we show the sparklines for the months from June 2018 to April 2022.
* If a range of 6+ months is selected, then in the UI show the sparklines for the selected range for weeks.
  * For example, if a date range of `2022-02-21,2022-10-19` is selected, then we show the sparklines for the week of `2022-02-21` to the week of  `2022-10-19`. 
* If a range of less than 6 months is selected, then we keep showing it as before.
* In the user interface, when hovering a sparkline, we should show a tooltip which data/period is shown in the sparkline graphs. Meaning which period and which dates. Like in the case of a selected date range `2022-02-21,2022-10-19` we would say something like "This sparklines shows the evolution over time from $periodStart to $periodEnd. Each data point represents a $period."
* When comparison different dates feature is used, then people might be comparing different ranges. This means we can maybe only convert to week/month if both the original range and the comparison range is of a certain length. When comparing only different periods then we can still easily use the same period.

Technical note:
We don't need to change the `Sparkline` visualisation itself but could simply change the [`Sparklines` visualisation](https://github.com/matomo-org/matomo/blob/4.13.0-rc2/plugins/CoreVisualizations/Visualizations/Sparklines.php#L224-L228) in above cases to not use `range` but `week` or `month` when needed. This way can then also more easily adjust the tooltip.

Example diff below (not taking into account yet the comparison feature)

```diff
diff --git a/plugins/CoreVisualizations/Visualizations/Sparklines.php b/plugins/CoreVisualizations/Visualizations/Sparklines.php
index 6ed99cea4f..e6e81c8979 100644
--- a/plugins/CoreVisualizations/Visualizations/Sparklines.php
+++ b/plugins/CoreVisualizations/Visualizations/Sparklines.php
@@ -13,6 +13,7 @@ namespace Piwik\Plugins\CoreVisualizations\Visualizations;
 use Piwik\API\Request;
 use Piwik\Common;
 use Piwik\DataTable;
+use Piwik\Date;
 use Piwik\Metrics;
 use Piwik\Metrics\Formatter as MetricFormatter;
 use Piwik\Period\Factory;
@@ -174,7 +175,8 @@ class Sparklines extends ViewDataTable
         $originalDate = Common::getRequestVar('date');
         $originalPeriod = Common::getRequestVar('period');
 
-        if ($this->isComparing() && !empty($comparisons)) {
+        $isComparing = $this->isComparing() && !empty($comparisons); 
+        if ($isComparing) {
             $comparisonRows = [];
             foreach ($comparisons->getRows() as $comparisonRow) {
                 $segment = $comparisonRow->getMetadata('compareSegment');
@@ -227,7 +229,23 @@ class Sparklines extends ViewDataTable
                 'action'  => $this->requestConfig->getApiMethodToRequest()
             ]);
 
-            if ($this->isComparing() && !empty($comparisons)) {
+            if ($originalPeriod === 'range' && !$isComparing) { // todo check for comparing segments it would still work to select different period but problem when comparing different periods
+                // when comparing, we have to fallback to a range because different periods might be selected.
+                // when not comparing and a longer range is selected, then we select a higher period for improved
+                // performance and also to see trends better
+                $periodObj = Factory::build($originalPeriod, $originalDate);
+                $numDaysInRange = $this->getNumDaysDifference($periodObj->getDateStart(), $periodObj->getDateEnd());
+                if ($numDaysInRange >= 730) {
+                    $sparklineUrlParams['period'] = 'month';
+                    $sparklineUrlParams['date'] = $originalDate;
+                } elseif ($numDaysInRange >= 180) {
+                    $sparklineUrlParams['period'] = 'week';
+                    $sparklineUrlParams['date'] = $originalDate;
+                }
+            }
+
+
+            if ($isComparing) {
                 $periodObj = Factory::build($originalPeriod, $originalDate);
 
                 $sparklineUrlParams['compareSegments'] = [];
@@ -285,6 +303,7 @@ class Sparklines extends ViewDataTable
                         'period' => $periodObj->getLabel(),
                         'date' => $periodObj->getLabel() === 'range' ? $periodObj->getRangeString() : $periodObj->getDateEnd(),
                     ]);
+
                     $this->config->addSparkline($params, $metrics, $desc = null, null, ($order * 100) + $segmentIndex, $title, $sparklineMetricIndex, $seriesIndices, $graphParams);
                 }
             } else {
@@ -322,6 +341,12 @@ class Sparklines extends ViewDataTable
         }
     }
 
+    private function getNumDaysDifference(Date $date1, Date $date2)
+    {
+        $days = (abs($date1->getTimestamp() - $date2->getTimestamp())) / 60 / 60 / 24;
+        return (int) round($days);
+    }
+
     private function applyFilters(DataTable\DataTableInterface $table)
     {
         foreach ($this->config->getPriorityFilters() as $filter) {
diff --git a/plugins/CoreVisualizations/Visualizations/Sparklines/Config.php b/plugins/CoreVisualizations/Visualizations/Sparklines/Config.php
index caf77a7b11..6cd87f6e7f 100644
--- a/plugins/CoreVisualizations/Visualizations/Sparklines/Config.php
+++ b/plugins/CoreVisualizations/Visualizations/Sparklines/Config.php
@@ -13,6 +13,7 @@ use Piwik\DataTable\Filter\CalculateEvolutionFilter;
 use Piwik\Metrics;
 use Piwik\NoAccessException;
 use Piwik\Period\Range;
+use Piwik\Piwik;
 use Piwik\Site;
 use Piwik\Url;
 
@@ -284,8 +285,15 @@ class Config extends \Piwik\ViewDataTable\Config
             $groupedMetrics[$metricGroup][] = $metricInfo;
         }
 
+        $tooltip = '';
+        if (!empty($requestParamsForSparkline['period'])) {
+            $tooltip = Piwik::translate('Each data point in the sparkline represents a %s.',
+                $requestParamsForSparkline['period'] === 'range' ? 'day' : $requestParamsForSparkline['period']);
+        }
+
         $sparkline = array(
             'url' => $this->getUrlSparkline($requestParamsForSparkline),
+            'tooltip' => $tooltip,
             'metrics' => $groupedMetrics,
             'order' => $this->getSparklineOrder($order),
             'title' => $title,
diff --git a/plugins/CoreVisualizations/templates/macros.twig b/plugins/CoreVisualizations/templates/macros.twig
index ea80c087bb..1c748b3bff 100644
--- a/plugins/CoreVisualizations/templates/macros.twig
+++ b/plugins/CoreVisualizations/templates/macros.twig
@@ -25,7 +25,9 @@
          {% if sparkline.seriesIndices|default is not empty %}data-series-indices="{{ sparkline.seriesIndices|json_encode|e('html_attr') }}"{% endif %}
          {% if sparkline.graphParams|default is not empty %}data-graph-params="{{ sparkline.graphParams|json_encode|e('html_attr') }}"{% endif %}
     >
-        <div>
+        <div
+            {% if sparkline.tooltip|default %}title="{{ sparkline.tooltip|e('html_attr') }}"{% endif %}
+        >
             {% if sparkline.title|default is not empty %}<h6 class="sparkline-title" title="{{ sparkline.title|rawSafeDecoded|e('html_attr') }}">{{ sparkline.title }}</h6>{% endif %}
             {% if sparkline.url %}{{ sparkline(sparkline.url)|raw }}{% endif %}
```

@mattab  can you check out this one?

### PR DESCRIPTION
**Pull request by tsteur**
### Description:

* fix #20066 when selecting currently eg a date range of 2 years it can take 2 minutes to load. Instead, we select 24 months which takes 2 seconds and allows the detection of trends much better. 
  * When now selecting a range of 6+ months Matomo will plot weeks in the sparklines
  * When now selecting a range of 24+ months Matomo will plot months in the sparklines
* Fix sparklines when comparing dates. Eg when you compared year vs previous year, then Matomo would load the previous 9 years in days (3000+days) and compared it to 1 year entry. It's completely broken visually and also performance wise it takes forever to load. Now no matter what period is selected, it will behave consistently and compare the selected dates. Eg when a year is compared to previous year, then we're showing the weeks of the first year and also plot the years of the second year so you can compare both years in detail. When comparing, selecting eg the previous 30 periods and comparing them is not as useful (although it can be). 
  * --> Right now without this change comparing any date of a week, month or year is basically broken as the sparklines never load.
  * --> It now also makes the sparklines behavior consistent with the evolution graph.
* We're now showing tooltips when possible to describe what date range and period the sparkline is showing. Before this was never clear eg when a year is selected, it's not obvious if Matomo is showing the months in that period or the previous years.
* Changed the logic to have things tested and clearly defined and so the behaviour can be changed easily.
* This PR does not change anything in the evolution graph.

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

File: plugins/CoreVisualizations/Visualizations/Sparklines/SparklinePeriodSelector.php [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+                $numDaysInRange = $this->getNumDaysDifference($period->getDateStart(), $period->getDateEnd());
+                if ($numDaysInRange < $lowestNumDaysInRange) {
+                    $lowestNumDaysInRange = $numDaysInRange;
+                }
+            }
+        }
+
+        $periodToUse = 'day';
+        if ($lowestNumDaysInRange >= 730) {
+            $periodToUse = 'month'; // rather than fetching > 730 day periods we prefer fetching 24+ months and shows trends better
```
Comment (bx80) [inline plugins/CoreVisualizations/Visualizations/Sparklines/SparklinePeriodSelector.php:115]: Would it be useful to have these values as configurable settings?

Comment (tsteur) [reply plugins/CoreVisualizations/Visualizations/Sparklines/SparklinePeriodSelector.php:115]: Thanks for checking @bx80 . Not needed in this case. We'd only add configs if really needed. Should a use case come up at some point then we can always add it later. 👍

**Review by bx80 — APPROVED** [review]
I've tested this on different date ranges and comparisons, the sparkline periods behave as expected with weeks being chosen over days for ranges over six months and months being chosen for ranges over two years. Should be a nice performance boost :+1:

* [x] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [x] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [x] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [x] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [x] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [x] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [x] Tests updated [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [x] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [x] n/a [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [x] n/a [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [x] n/a Existing documentation updated if needed

**Comment by tsteur** [comment]
Thanks @bx80 for reviewing 👍

**Pull request merged**